include/pybind11/pytypes.h

```diff
@@ -1336,9 +1337,14 @@ class memoryview : public object {
         // Py_buffer uses signed sizes, strides and shape!..
         static std::vector<Py_ssize_t> py_strides { };
         static std::vector<Py_ssize_t> py_shape { };
+        static std::vector<std::string> formats = {
```

Comment:
This is not future proof: if codes are added in future Python versions, this function will fail in highly unobvious and confusing ways.
https://github.com/python/cpython/blob/master/Objects/memoryobject.c#L1162
I see formats, after curation there, can have two characters. Is that a concern here?

Reply:
Understood. This should perhaps support unknown format that may appear in future. Currently there seems only `@` can appear in the format as a prefix, so the length is at most 2.

The main problem here is that `const char*` must reside after `py::buffer_info` goes out of scope, and this constructor needs to take care of this. As cpython's `get_native_fmtstr ` is not public, there needs a list of supported formats somewhere inside `pybind11`. Naive approach is to have a static set of strings that caches a given string, but this won't have any format validation.

Relevant: https://bugs.python.org/issue15821